docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -102,6 +102,289 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
+        }
+
+        return new HivePartitionResult(partitionColumns, partitionNames, partitionsIterable, effectivePredicate, compactEffectivePredicate, hiveBucketHandle, bucketFilter);
+    }
+
+First, the ``TupleDomain`` from the constraint is intersected with the
+``TupleDomain`` already applied to the ``TableHandle`` to form an
+``effectivePredicate``. If the result is not empty, it can be used as
+a filter on Hive buckets and partitions. Note that we're using ``predicate``
+only for partition filtering.
+
+Now let's take a look at how a new ``HiveTableHandle`` is created from
+an old one and an additional constraint:
+
+.. code-block:: java
+
+    public HiveTableHandle applyPartitionResult(HiveTableHandle handle, HivePartitionResult partitions, Constraint constraint)
+    {
+        Optional<List<String>> partitionNames = partitions.getPartitionNames();
+        Optional<List<HivePartition>> partitionList = Optional.empty();
+        TupleDomain<ColumnHandle> enforcedConstraint = handle.getEnforcedConstraint();
+
+        // Partitions will be loaded if
+        // 1. Number of partitionNames is less than or equal to threshold value. Thereby generating additional filter criteria
+        //    that can be applied on other join side (if the join is based on partition column),
+        // 2. If additional predicate is passed as a part of Constraint. (specified via loadPartition). This delays the partition checks
+        //    until we have additional filtering based on Constraint
+        if (canPartitionsBeLoaded(partitions) || constraint.predicate().isPresent()) {
+            partitionNames = Optional.empty();
+            partitionList = Optional.of(getPartitionsAsList(partitions));
+            List<HiveColumnHandle> partitionColumns = partitions.getPartitionColumns();
+            enforcedConstraint = partitions.getEffectivePredicate().filter((column, domain) -> partitionColumns.contains(column));
+        }
+        return new HiveTableHandle(
+                handle.getSchemaName(),
+                handle.getTableName(),
+                handle.getTableParameters(),
+                ImmutableList.copyOf(partitions.getPartitionColumns()),
+                handle.getDataColumns(),
+                partitionNames,
+                partitionList,
+                partitions.getCompactEffectivePredicate(),
+                enforcedConstraint,
+                partitions.getBucketHandle(),
+                partitions.getBucketFilter(),
+                handle.getAnalyzePartitionValues(),
+                Sets.union(handle.getConstraintColumns(), constraint.getPredicateColumns().orElseGet(ImmutableSet::of)),
+                handle.getProjectedColumns(),
+                handle.getTransaction(),
+                handle.isRecordScannedFiles(),
+                handle.getMaxScannedFileSize());
+    }
+
+To see how to handle expressions in WHERE constraint pushdown, let's take
+a look at the base JDBC connector:
```

Comment:
I don't like how we suddenly switch from one pretty complex example to another, complex one. Why couldn't we have started with JDBC?

Reply:
Would it be better to start with an example where a connector only knows how to push down a single value predicate on a specific column and ignores everything else? We could introduce all the necessary data structures this way, building up confidence in the reader to move on to more advanced topics.

Reply:
> I don't like how we suddenly switch from one pretty complex example to another, complex one. Why couldn't we have started with JDBC?

JDBC doesn't look at the predicate, and I wanted to show that the predicate has priority over the summary.

Reply:
> > I don't like how we suddenly switch from one pretty complex example to another, complex one. Why couldn't we have started with JDBC?
> 
> JDBC doesn't look at the predicate, and I wanted to show that the predicate has priority over the summary.
> 
> > Would it be better to start with an example where a connector only knows how to push down a single value predicate on a specific column and ignores everything else? We could introduce all the necessary data structures this way, building up confidence in the reader to move on to more advanced topics.
> 
> Possibly, but the connector would need to exist outside of the docs, otherwise it would be difficult to navigate. Maybe we could only show how to use a `TupleDomain` and point the reader to existing connectors to let them figure out expression and predicate handling?

My point is to have a complete example that readers can follow, instead of having to figure out things from complete connectors, where there can be a lot of code not relevant to what they need to do. We don't need to have a complete working connector to show how to handle a single predicate.